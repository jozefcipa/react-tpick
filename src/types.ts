
export interface Props {
  hours: number
  minutes: number
  hoursStep?: number
  minutesStep?: number
  onChange: ({ hours, minutes }: { hours: number, minutes: number }) => void
}

export interface ChangeValueParams {
  minutesValue?: number
  hoursValue?: number
  step: number
}
