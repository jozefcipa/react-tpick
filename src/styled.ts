import styled from 'styled-components'

// import * as upCaret from './assets/up-caret.svg'
// import * as downCaret from './assets/up-caret.svg'

export const Container = styled.div`
  width: 50px;
  display: flex;
  justify-content: center;
  align-items: center;
`

export const Input = styled.input`
  width: 20px;
  border: none;
  text-align: center;
  margin: .5rem;
  font-size: 1rem;
  &:focus {
    outline: none;
  }
`

export const Switch = styled.div`
  display: flex;
  flex-direction: column;
  justify-content: center;
`

const Button = styled.span`
  align-self: stretch;
  background-repeat: no-repeat;
  background-size: 10px;
  background-position: center;
  height: 10px;
  cursor: pointer;
`

export const UpButton = styled(Button)`
    background-image: url(${'upCaret'});    
`

export const DownButton = styled(Button)`
    background-image: url(${'downCaret'}); 
`
