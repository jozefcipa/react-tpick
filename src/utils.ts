export const timeStringValue = (val: number): string => val < 10 ? `0${val}` : `${val}`

export const getIncreasedValueOrReset = (val: number, maxValue: number): number => {
  if (val > maxValue) {
    return 0
  }
  return val
}

export const checkHoursStep = (step: number, max: number): void => {
  if (step > max || step < 1) {
    throw Error('Hours step has to be in range 1-23')
  }
}

export const checkMinutesStep = (step: number, max: number): void => {
  if (step > max || step < 1) {
    throw Error('Minutes step has to be in range 1-59')
  }
}
