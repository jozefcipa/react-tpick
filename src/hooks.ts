import * as React from 'react'
import { Time, UpdateTimeParams, UpdateTimeSegment, UseTimeHook } from './types'
import { checkHoursStep, checkMinutesStep, getIncreasedValueOrReset } from './utils'

const MAX_HOURS = 23
const MAX_MINUTES = 59

export const useTime = (initValue: Time): UseTimeHook => {
  const [hours, setHours] = React.useState(initValue.hours)
  const [minutes, setMinutes] = React.useState(initValue.minutes)

  const setTime = (time: Time): void => {
    setHours(time.hours)
    setMinutes(time.minutes)
  }

  const increment = (updateObj: UpdateTimeParams): Time => {
    const newTime = { hours, minutes }

    // increment hours
    if (updateObj.update === UpdateTimeSegment.Hours) {
      checkHoursStep(updateObj.step, MAX_HOURS)
      const newHours = hours + updateObj.step
      newTime.hours = getIncreasedValueOrReset(newHours, MAX_HOURS)
    }

    // increment minutes
    if (updateObj.update === UpdateTimeSegment.Minutes) {
      checkMinutesStep(updateObj.step, MAX_MINUTES)
      const newMinutes = minutes + updateObj.step
      if (newMinutes > MAX_MINUTES) {
        newTime.minutes = 0
        // TODO: increment hours
        // if (true) { // TODO: come up with some prop name
        //   newTime.hours = getIncreasedValueOrReset(newTime.hours + 1, MAX_HOURS)
        // }
      } else {
        newTime.minutes = newMinutes
      }
    }
    return newTime
  }

  const decrement = (updateObj: UpdateTimeParams): Time => {
    const newTime = { hours, minutes }

    // decrement hours
    if (updateObj.update === UpdateTimeSegment.Hours) {
      checkHoursStep(updateObj.step, MAX_HOURS)
      const newHours = hours - updateObj.step
      if (newHours < 0) {
        newTime.hours = MAX_HOURS
      } else {
        newTime.hours = newHours
      }
    }

    // decrement minutes
    if (updateObj.update === UpdateTimeSegment.Minutes) {
      checkMinutesStep(updateObj.step, MAX_MINUTES)
      const newMinutes = minutes - updateObj.step
      if (newMinutes < 0) {
      // TODO: decrement hours
        newTime.minutes = (MAX_MINUTES + 1) - updateObj.step
      } else {
        newTime.minutes = newMinutes
      }
    }
    return newTime
  }

  return [
    { hours, minutes },
    { increment, decrement, setTime },
  ]
}
