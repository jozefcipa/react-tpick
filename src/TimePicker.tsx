import * as React from 'react'
import { useTime } from './hooks'
import { Switch, Container, Input, UpButton, DownButton } from './styled'
import { timeStringValue } from './utils'
import { Props, UpdateTimeSegment } from './types'

const TimePicker: React.FC<Props> = (props: Props) => {
  const [time, actions] = useTime({
    hours: props.hours,
    minutes: props.minutes,
  })

  // step
  const [hoursStep, setHoursStep] = React.useState(props.hoursStep || 1)
  const [minutesStep, setMinutesStep] = React.useState(props.minutesStep || 1)

  // update step timeStringValue
  React.useEffect(() => {
    setHoursStep(props.hoursStep || 1)
    setMinutesStep(props.minutesStep || 1)
  }, [props.hoursStep, props.minutesStep])

  React.useEffect(() => {
    actions.setTime({
      hours: props.hours,
      minutes: props.minutes,
    })
  }, [props]) // eslint-disable-line react-hooks/exhaustive-deps

  return (
    <Container>
      <Switch>
        <UpButton
          onClick={() => props.onChange(actions.increment({ update: UpdateTimeSegment.Hours, step: hoursStep }))}
        >
          +
        </UpButton>
        <Input type="text" value={timeStringValue(time.hours)} disabled />
        <DownButton
          onClick={() => props.onChange(actions.decrement({ update: UpdateTimeSegment.Hours, step: hoursStep }))}
        >
          -
        </DownButton>
      </Switch>
      <span>:</span>
      <Switch>
        <UpButton
          onClick={() => props.onChange(actions.increment({ update: UpdateTimeSegment.Minutes, step: minutesStep }))}
        >
          +
        </UpButton>
        <Input type="text" value={timeStringValue(time.minutes)} disabled />
        <DownButton
          onClick={() => props.onChange(actions.decrement({ update: UpdateTimeSegment.Minutes, step: minutesStep }))}
        >
          -
        </DownButton>
      </Switch>
    </Container>
  )
}

export default TimePicker
