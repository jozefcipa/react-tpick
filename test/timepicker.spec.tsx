import * as React from 'react'
import * as renderer from 'react-test-renderer'
import { TimePicker } from '../src'

it('renders correctly', () => {
  const tree = renderer.create(<TimePicker hours={10} minutes={45} onChange={() => {}}/>).toJSON()
  expect(tree).toMatchSnapshot()
})
