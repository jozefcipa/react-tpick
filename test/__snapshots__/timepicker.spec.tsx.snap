// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="sc-bwzfXH hSZadk"
>
  <div
    className="sc-bxivhb dzPXKO"
  >
    <span
      className="sc-bdVaJa sc-ifAKCX QmPtq"
      onClick={[Function]}
    />
    <input
      className="sc-htpNat cmIQeZ"
      disabled={true}
      type="text"
      value="10"
    />
    <span
      className="sc-bdVaJa sc-EHOje exncpR"
      onClick={[Function]}
    />
  </div>
  <span>
    :
  </span>
  <div
    className="sc-bxivhb dzPXKO"
  >
    <span
      className="sc-bdVaJa sc-ifAKCX QmPtq"
      onClick={[Function]}
    />
    <input
      className="sc-htpNat cmIQeZ"
      disabled={true}
      type="text"
      value="45"
    />
    <span
      className="sc-bdVaJa sc-EHOje exncpR"
      onClick={[Function]}
    />
  </div>
</div>
`;
